src/main/host/descriptor/pipe.rs

```diff
@@ -91,10 +97,13 @@ impl PipeFile {
             return Err(nix::errno::Errno::EBADF.into());
         }
 
-        let num_written = self.buffer.borrow_mut().write(bytes, event_queue);
+        let num_written = self
+            .buffer
+            .borrow_mut()
+            .write(bytes.by_ref(), event_queue)?;
 
         // the write would block if we could not write any bytes, but were asked to
-        if num_written == 0 && !bytes.is_empty() {
+        if usize::from(num_written) == 0 && bytes.bytes().next().is_some() {
             Err(nix::errno::EWOULDBLOCK.into())
```

Comment:
> A Take object also implements Read. If the caller wants to put a cap on how much data is read into the ByteQueue, they can already use Take to do so. We do this a few lines up from here.

The caller can use `take()` to limit how much data, but if the `PipeFile::write()` accepts only a `Read`, then it loses that information (it can't know the limit without reaching that limit).

> Unfortunately I don't think there is such an adapter in the standard library for Write. I don't think we'll need it in practice though, and if we do, I think it'd be cleaner to find or create such an adapter (Give?) .

Looking at this again, I think I left a bug in `PipeFile::read()`, where it's not returning a `EWOULDBLOCK` instead of 0 if the pipe is empty :( So I think we will also need to know if the Write is non-empty or not.

> In an earlier iteration ProcessMemoryReader and ProcessMemoryWriter also implemented the Seek trait, and I required the Seek trait for R here. Then here we can check whether seek(SeekFrom::Start(1)).is_ok(), and then rewind it before returning via seek(SeekFrom::Start(0)). I think I'll probably put that back in, but am open to other ideas.

I've been trying to think of an alternative to Seek, but I can't think of any.

Reply:
Done. This cascaded a bit - it seemed silly to implement Seek twice, so I extracted that and some other pointer arithmetic into `TypedPluginPtr`.

I ended up backporting std::io::Seek::stream_len. Not sure this is the best way to do it. Do you know if we switched to nightly, would we only get nightly-only features that we explicitly opt into? If that's the case, maybe it'd be worth thinking about.